### PR TITLE
WIP: new loader.py, proof of concept

### PR DESCRIPTION
**Pull request by ninja-**
### What does this PR do?
POC implementation of #39011.
Currently module loading is slow, because it's loading random modules and these modules try to load other modules etc.

Seperated gathering module virtualnames and the part that calls __virtual__ so that it will be called only when module is directly requested.

I am calling it POC because I am not able to show a direct performance improvement _yet_.
I am thinking about implementing some on-disk cache here. :)

**Also, it needs to load the modules into classpath which itself takes time
(haven't measured yet how much...)**
But that's something that would be solved with on-disk cache here!

EDIT: also, I think there's a logical bug left somewhere because I was able to use all the modules but one failed(grains?). Leaving a note so I won't forget to check. I don't remember exactly right now.

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

File: salt/loader.py [hunk 1]
```diff
@@ -1062,7 +1064,8 @@ def __init__(self,
         # names of modules that we don't have (errors, __virtual__, etc.)
         self.missing_modules = {}  # mapping of name -> error
         self.loaded_modules = {}  # mapping of module_name -> dict_of_functions
-        self.loaded_files = set()  # TODO: just remove them from file_mapping?
+        self.loaded_files = {}  # TODO: just remove them from file_mapping?
```
Comment (mirceaulinic) [inline salt/loader.py]: Jus out of curiosity: is `{}` faster than `set()`? I know there's a performance difference between `[]` and `list()` for example.

However, I looking at that line, I would understand that `self.loaded_files` is a dictionary. So please don't forget to add a comment to specify that is a set.

Comment (ninja-) [inline salt/loader.py]: It's now mapping file names to the module objects, so set/array isn't enough and dict is used.

Comment (mirceaulinic) [inline salt/loader.py]: Oh, my bad. Please disregard my comments.

**Comment by thatch45** [comment]
I don't think that this will speed up module loads, I am also worried about breaking the process of dep lookups like this. The loader is somewhat complicated and works well, I know it is not really fast, but it does cache loaded modules and references, so it is only a little slow on the first run.
Basically, I need to say that we should not merge this, without proof or a significant performance improvement.
It is also important to point out that while the loader is not really fast, it is not the bottleneck that we typically see in Salt, I think we would get a lot more bang for our buck by focusing on areas that are causing more regular performance issues

**Comment by ninja-** [comment]
> I don't think that this will speed up module loads, I am also worried about breaking the process of dep lookups like this.

It's not faster yet. The dependency lookups are going to be fine, trust me :)


> The loader is somewhat complicated and works well, I know it is not really fast, but it does cache loaded modules and references, so it is only a little slow on the first run.

Unfortunately that's half-truth. It needs to load them every time you use salt-call and it's probably same via master-minion setup. That takes significant time.

> Basically, I need to say that we should not merge this, without proof or a significant performance improvement.

You are right. That's why I mentioned the on-disk cache of virtual names mapping, which I would finish soon, to provide you with final performance results from a real-life setup to evaluate.

> It is also important to point out that while the loader is not really fast, it is not the bottleneck that we typically see in Salt, I think we would get a lot more bang for our buck by focusing on areas that are causing more regular performance issues

I am in process of optimizing my slow saltstack deploy time and I am focusing on several areas:
1. Loader being slow, eating time every time you call salt
2. Jinja templates for several formulas being slow (i.e. PHP) - 0.5s per file x 5 templates for example
3. The reload_modules bug and other performance bugs all over SaltStack
4. Slow up-to-date check for file templates for file.manages
So file.managed via salt-call is 5ms and 40ms via master
5. Pillar rendering being slow sometimes in my setup

All of that things are adding to the "overhead" which is the difference between total process time of salt-call and "total states run" time.

**Comment by thatch45** [comment]
Sounds good. I just want to be clear that we need to be very careful in here and I will not promise that we will merge any of this unless we are confident that theses systems will stay stable. The last time we modified the loader for performance we were finding bugs for quite some time and we are less then excited about going through that again :)

**Comment by ninja-** [comment]
Note: I now have a much better idea on how to optimize the loader without breaking stuff.

How I want it to work:
For each module dir (primarly the system path = /usr/lib/python3/site-module/salt/{modules,states})
it's going to generate an index file in a following way:
- load all modules to the classpath, without calling __virtual__
(some poorly written modules might crash at this point already, because they are importing some dependencies outside __virtual__, so I need to handle that too)
- map all virtualnames and the function names
- index generation should take around ~ 5 seconds
- the index is going to be generated only once, and it will be renewed when last-modified of /usr/lib/python3/site-modules/salt is greater than last index time


We could make the index generation part of salt's build process, so the package would ship with something like modules.map

I'll try to implement that soon enough. Thoughts?
😄

**Comment by cachedout** [comment]
@thatch45 Could you swing by this one and comment on the above, please?

**Comment by thatch45** [comment]
I really like that idea, my worry is in calling virtual, sure you are not SUPPOSED to put critical stuff in virtual, but it happens, and if we have a pre index of modules to load then we will load the right module first, so calling virtual should not be a bid deal.
All in all, I would say, make sure it still falls back to to scan for the module if the indexed module for some reason does not return the virtual you need and to still double check the virtual.

**Comment by rallytime** [comment]
Since we haven't seen any updates here for a while, I am going to close this. We can always reopen when this is ready to go and continue the discussion. Thanks everyone!

**Pull request closed without merging**